Initial Kafka messages parsing

avelanarius opened the pull request:
Implemented parsing of basic Kafka protocol primitives, as well as three messages: `ApiVersions`, `Metadata`, `Produce` requests and responses.

The parsing of records (`kafka_records.cc`) is not fully feature complete, as it lacks support of compressions, versions 0 or 1 and validation of CRC value.

avelanarius committed:
kafka: Primitives, ApiVersions, Metadata parsing

Implement classes for parsing basic Kafka protocol primitives,
as well as ApiVersions and Metadata request/response messages.

StarostaGit commented on tests/unit/CMakeLists.txt:
```diff
@@ -309,6 +309,10 @@ seastar_add_test (ipv6
 seastar_add_test (json_formatter
   SOURCES json_formatter_test.cc)
 
+seastar_add_test (kafka_protocol_test
```
NIT: I think the naming convention for adding test files "name_test.cc" is to leave only "name" in cmake name, i.e. `seastar_add_test (name ...)`, therefore:
```suggestion
seastar_add_test (kafka_protocol
```

constantfold replied:
Fixed in: https://github.com/psarna/seastar/pull/17/commits/55034090d38cc1c5567fddea825364d2859add57#diff-90e79c2c9c9258ee3a1bbac766e6c402R312

avelanarius committed:
kafka: Varint parsing

Implement 32-bit zigzag varint parsing, as defined in the Kafka
protocol specification.

avelanarius committed:
kafka: RECORDS type structures and record parsing

Implement structures storing RECORDS type (as defined in Kafka protocol
specification) and parsing of a single record (only version 2).

avelanarius committed:
kafka: Parsing of records batch

Implement parsing of records batch from RECORDS Kafka protocol
type (version 2).

avelanarius committed:
kafka: Parsing of RECORDS type

Implemented parsing of RECORDS type (version 2), as defined in the
Kafka protocol specification.

avelanarius committed:
kafka: Produce request parsing

Implement parsing of Produce request from Kafka protocol.

avelanarius committed:
kafka: Produce response parsing

Implement parsing of Produce response from Kafka protocol.

avelanarius commented:
Question for reviewers (@haaawk , @psarna): Would it be OK to get rid of getters and setters (make fields public) in message classes, as getters/setters makes filling those messages very cumbersome (as they are nested) and there is no validation/extra functionality in those setters/getters (probably still true in the future).

psarna commented:
@avelanarius totally fine, especially during early development, this isn't Java :) Validation can be provided later with free functions.

avelanarius commented:
Pushed the commit to remove getters/setters (https://github.com/psarna/seastar/pull/17/commits/c221f3a964108ed5d6d63efe152628132b0ecb2c). Hopefully it will also make code review a bit easier :)

@psarna @haaawk I'd be grateful if you could review this PR before Friday (or tommorow), as the Monday deadline is very soon (and there is next PR with producer + demo waiting for this PR to be merged - of course after neccessary fixes if they are recommended by you).

psarna commented:
If this code is considered ready, please squash and merge this series, so that it doesn't contain fixes to itself - you can leverage `git rebase -i` for that purpose. In Seastar/Scylla, a branch is only reviewed and merged once it has a clear history - in particular, all commits that do cleanup and fixes of other commits in the same series should be merged into commits they fix. That applies to (at least) 482c7dd8ccc48bbe31ecceb658cb34d2015b3125 55034090d38cc1c5567fddea825364d2859add57  and c221f3a964108ed5d6d63efe152628132b0ecb2c.

We need that because a clear history simplifies backporting, rebasing and reviewing a lot, it also makes it way easier to read git log. Fixups are totally fine during development, but they need to be squashed before the final review.

avelanarius commented:
@psarna I did the rebase and squashed fixup commits. Decided not to merge multiple commits regarding Records and Produce parsing, as they are really implementing separate funtionality (they are not fixups).

> If this code is considered ready

I consider my code ready.

psarna commented on src/kafka/protocol/kafka_primitives.hh:
```diff
@@ -0,0 +1,321 @@
+    void deserialize(std::istream &is, int16_t api_version) {
+        kafka_number_t<SizeType> length;
+        length.deserialize(is, api_version);
+        if (*length < 0) {
```
does Kafka define max length? It would be better to check rather than accept a really large number here and crash the whole server with `std::bad_alloc`. You can also leave a TODO.

constantfold replied:
It seems that there is no upper bound on length in protocol specification. Some messages (Fetch) allow you to specify max_bytes. Adding a TODO for now.

psarna commented on src/kafka/protocol/kafka_primitives.hh:
```diff
@@ -0,0 +1,321 @@
+    }
+
+    [[nodiscard]] std::string &operator*() {
+        if (_is_null) throw std::domain_error("Object is null.");
```
we have a strict policy of always putting curly braces around `if` statements:
```cpp
if (...) {
    (...)
}
```

psarna commented on src/kafka/protocol/kafka_primitives.hh:
```diff
@@ -0,0 +1,321 @@
+    }
+
+    [[nodiscard]] const std::string *operator->() const {
+        if (_is_null) throw std::domain_error("Object is null.");
```
ditto, applies to all `if` statements below as well

psarna commented on src/kafka/protocol/kafka_primitives.hh:
```diff
@@ -0,0 +1,321 @@
+    void deserialize(std::istream &is, int16_t api_version) {
+        kafka_number_t<SizeType> length;
+        length.deserialize(is, api_version);
+        if (*length >= 0) {
```
again, a verification of max length would be nice here (or, a TODO)

psarna reviewed (commented):
Generally looks good, I had minor comments. As for the code style, everything is linked here: https://github.com/scylladb/scylla/wiki/Formatting-and-sending-patches (along with how commit messages should be formatted - in particular - lines are too long).

@haaawk once my issues are addressed, feel free to merge if it looks good to you as well

psarna commented:
By the way, @avelanarius, it's totally fine to create pull requests on top of pull requests (just by checking out someone else's code and working on top of it), so even if this one sits in review for a while, it should not block other PRs or demos - it's fine just to mention in the issue that a particular PR depends on the other one. Git offers powerful tools for rebasing, cherry-picking and merging, so it will not be hard to merge all the PRs, even if some of them depend on other ones.

avelanarius commented:
Fixed the issues (max length todos, ifs with brackets). 
Fixed the formatting of commits.
The branch is also rebased on top of latest `zpp_kafka`.

@haaawk Please review and merge if there are no new issues.

haaawk commented:
LGTM in general but I left few comments
